pennylane/devices/qubit_mixed/measure.py

```diff
@@ -0,0 +1,289 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code relevant for performing measurements on a qutrit mixed state.
+"""
+
+from collections.abc import Callable
+
+from pennylane import math, queuing
+from pennylane.measurements import (
+    ExpectationMP,
+    MeasurementProcess,
+    ProbabilityMP,
+    StateMeasurement,
+    StateMP,
+    VarianceMP,
+)
+from pennylane.ops import Sum
+from pennylane.typing import TensorLike
+from pennylane.wires import Wires
+
+from .apply_operation import _get_num_wires, apply_operation
+
+
+def _reshape_state_as_matrix(state, num_wires):
+    """Given a non-flat, potentially batched state, flatten it to square matrix or matrices if batched.
+
+    Args:
+        state (TensorLike): A state that needs to be reshaped to a square matrix or matrices if batched
+        num_wires (int): The number of wires the state represents
+
+    Returns:
+        Tensorlike: A reshaped, square state, with an extra batch dimension if necessary
+    """
+    dim = 2**num_wires
+    batch_size = math.get_batch_size(state, ((2,) * (num_wires * 2)), dim**2)
+    shape = (batch_size, dim, dim) if batch_size is not None else (dim, dim)
+    return math.reshape(state, shape)
+
+
+def calculate_expval(
+    measurementprocess: ExpectationMP,
+    state: TensorLike,
+    is_state_batched: bool = False,
+    readout_errors: list[Callable] = None,
+) -> TensorLike:
+    """Measure the expectation value of an observable.
+
+    Args:
+        measurementprocess (ExpectationMP): measurement process to apply to the state.
+        state (TensorLike): the state to measure.
+        is_state_batched (bool): whether the state is batched or not.
+        readout_errors (List[Callable]): List of chanels to apply to each wire being measured
+        to simulate readout errors.
+
+    Returns:
+        TensorLike: expectation value of observable wrt the state.
+    """
+    probs = calculate_probability(measurementprocess, state, is_state_batched, readout_errors)
+    eigvals = math.asarray(measurementprocess.eigvals(), dtype="float64")
+    # In case of broadcasting, `probs` has two axes and these are a matrix-vector products
+    return math.dot(probs, eigvals)
+
+
+def calculate_reduced_density_matrix(
+    measurementprocess: StateMeasurement,
+    state: TensorLike,
+    is_state_batched: bool = False,
+    _readout_errors: list[Callable] = None,
```

Comment:
I'd rather that we kept the function signatures consistent.

```suggestion
    readout_errors: list[Callable] = None,
```